### ISSUE TITLE
Less dependency version expected?

### ISSUE THREAD

**Issue by sraedler** [VeNoMouS/cloudscraper#91]
First, thanks a lot for your awesome and funny work!

I just Wanted to ask, whether it is possible to create a interpreter with less dependencies, especially without the need of pip?

Why I‘m asking?
Because I want to use the addon with Kodi and there is no pip. I know it is possible to add it as other modules, but that‘s an additional dependency.

I already saw some modules with a kind of hardcoded js evaluation.

Do you think this is possible?

Thanks in advance!

**Comment by VeNoMouS** [comment]
Yea, there are alot of areas I need to clean up that haven't been touched in months...

```I already saw some modules with a kind of hardcoded js evaluation.```

^^ What do you mean?

re kodi packaging... how are you going to deal with the interpreter dependency, are you going to have the user install nodejs for example themselves? I see there is a 3rd party js2py kodi module so you could use that...


Either way i'll put the dependency stuff on my todo

(btw ive written my own kodi modules too ;P)

**Comment by sraedler** [comment]
I tried it with js2py as modul.
But I don‘t like all the dependencies...

Would you like to share your work for Kodi? ;)

With hardcoded I mean something like that: https://github.com/a4k-openproject/script.module.openscrapers/blob/master/lib/openscrapers/modules/cfscrape_solver.py

**Comment by VeNoMouS** [comment]
^^^ just looking at that module... i dont see how that would work... unless __ord()__ is doing jsunfuck... but i highly doubt it...  it looks like a varition of @doko-desuka && @Arias800 's work... in the past I tried to avoid the whole custom python js solver... but if i find some time I could look into it again.

**Comment by sraedler** [comment]
In the same git folder is also a jsunfuck file. Maybe that helps you.

I hope you can solve this :)

BTW. Kodi can also not handle OpenSSL. Since it is only used to catch the exception, I think it might be replaced with an buildin modul/exception?

**Comment by doko-desuka** [comment]
@sraedler in what way is the OpenScrapers' version, or [the version in Arias800](https://github.com/Arias800/CloudflareScraper/blob/master/CloudflareScraper/__init__.py), not working for you in Kodi? If you find the solution to this it would spare @VeNoMouS's time to work on other things.

I use Arias800 on a personal video add-on, it works fine. From time to time it gives a wrong answer, but I've not bothered to debug it because it's rare and asking for the page again gives a different challenge that is then successfully solved.

**Comment by VeNoMouS** [comment]
@doko-desuka Cloudflare changed their whole challenge delivery yesterday, to POST and new vars, js challenge payload still the same

**Comment by doko-desuka** [comment]
Oh sh!t. Good to know, thank you.

**Comment by VeNoMouS** [comment]
Prob best to look at my latest code.. since i've made improvements since that commit...

**Comment by doko-desuka** [comment]
I did, it helped a lot.

**Comment by sraedler** [comment]
@doko-desuka kinoger.com for example does not work.

Do you two see a way to combine your work, since @VeNoMouS fixes the ciphers nearly daily.

**Comment by doko-desuka** [comment]
Ours only works on that one if you try it first with the HTTP scheme to get the cookie, then it redirects to HTTPS.  
It's probably because of the SSL ciphers (picking a user-agent at random, without matching it with the proper cipher suite, it's using whatever urllib3 has).

**Comment by doko-desuka** [comment]
Okay, thanks to VeNoMouS' very useful code and this doc section (https://www.npmjs.com/package/cloudscraper#recaptcha), I was able to add an adapter to that CF solver for Kodi. Tests were succesful.

The adapter is this. Note that it does not check any certification or hostname, because for video streaming we don't need this, we're just consuming data instead of submitting data, so it doesn't matter if the site is hijacked as long as the stream is good:
```
import requests

(...)

class CloudflareAdapter(requests.adapters.HTTPAdapter):
    # Cloudflare-friendly HTTP adapter based on VeNoMouS' Cloudscraper.
    # (1: https://github.com/VeNoMouS/cloudscraper/)
    # (2: https://www.npmjs.com/package/cloudscraper#recaptcha)
    def __init__(self, **kwargs):
        self.ssl_context = requests.packages.urllib3.util.ssl_.create_urllib3_context(
            ssl_version = ssl.PROTOCOL_SSLv23, # SSLv2, SSLv3, TLSv1, TLSv1.1, TLSv1.2
            cert_reqs = ssl.CERT_NONE,
            options = None, # Defaults to ssl.OP_NO_SSLv2, ssl.OP_NO_SSLv3 and ssl.OP_NO_COMPRESSION.
            ciphers = 'GREASE-2A2A:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA384:ECDHE+AESGCM:'
            'ECDHE+CHACHA20:DHE+AESGCM:DHE+CHACHA20:ECDH+AESGCM:DH+AESGCM:ECDH+AES:DH+AES:RSA+AESGCM:RSA+AES:'
            '!aNULL:!eNULL:!MD5:!DSS'
        )
        self.ssl_context.check_hostname = False
        super(CloudflareAdapter, self).__init__(**kwargs)

    def init_poolmanager(self, *args, **kwargs):
        kwargs['ssl_context'] = self.ssl_context
        return super(CloudflareAdapter, self).init_poolmanager(*args, **kwargs)

    def proxy_manager_for(self, *args, **kwargs):
        kwargs['ssl_context'] = self.ssl_context
        return super(CloudflareAdapter, self).proxy_manager_for(*args, **kwargs)
```

To be used with `mySession.mount('https://', CloudflareAdapter())`, preferably in the \_\_init\_\_() method of your requests.Session subclass.
I'm also spoofing the headers from my Nexus 7 tablet. You can get the headers of your device(s) by using sites like [CyLog](https://www.cylog.org/headers/) or [HTTPBin](https://httpbin.org/get), and then [print the page with Chrome mobile](https://support.google.com/chrome/answer/1069693?co=GENIE.Platform%3DAndroid&oco=1) or whatever, send it to yourself by email etc.

**Comment by VeNoMouS** [comment]
FWIW "grease" is a boringSSL thing, that isnt actually the real cipher name...  https://www.chromestatus.com/feature/6475903378915328

**Comment by doko-desuka** [comment]
It came up on the SSL inspection, and I wanted to copy it as much as possible:
![nexus7Ciphers](https://user-images.githubusercontent.com/31810999/69869425-981ccd80-128b-11ea-8484-1c7bf05b4def.png)
The other ciphers like "CHACHA20" etc I have no idea what aliases could be used.  
I don't know if this makes any practical difference to be honest. If it doesn't then I'd better remove it.

**Comment by VeNoMouS** [comment]
@doko-desuka refer to my browser.json in User_Agent dir ;P ciphers for chrome / firefox are in there...

or check the dev tree for a more "generic" one for custom user-agent's....

**Comment by doko-desuka** [comment]
I saw that, did you make that on your own? I don't think there's any other (open) database like that.

**Comment by VeNoMouS** [comment]
Yea I did

**Comment by doko-desuka** [comment]
> the prefered method is ssl.PROTOCOL_TLS

I used the table in "Socket Creation", they should've updated it with that info.
https://docs.python.org/2/library/ssl.html#socket-creation

**Comment by VeNoMouS** [comment]
yeaaaaaa take that SSL ordering with a grain of salt in python... it doesn't follow the order as far as i've seen...

**Comment by VeNoMouS** [comment]
The number one issue i've found with cloudflare... for no reason at all... even though say chrome or firefox is presenting the ciphers normally..... they will randomly decide to start reCaptching if a cipher is in there...

I need to spend some more time messing around with cloudflare every ~6+ days decided to randomly block a cipher.. yet the browsers arent affected by it...

**Comment by doko-desuka** [comment]
Okay, if I find any problems I'll report back, just for reference.  
The Chrome in my Nexus also sends these modern headers, I don't know if they help with spoofing as a mobile device:
```
            ('Sec-Fetch-User', '?1'),
            ('Sec-Fetch-Site', 'same-origin'),
            ('Sec-Fetch-Mode', 'navigate'),
```

**Comment by VeNoMouS** [comment]
@doko-desuka nah they not needed...

you can select mobiles via my user_agent class

ie

scraper = cloudscraper.create_scraper(browser={'browser': 'chrome', 'desktop': False})

or... 'mobile': False... if you want only desktop versions etc...

i added in "custom" today into the dev branch with a "generic" cipherSuite...

**Comment by VeNoMouS** [comment]
anyway @doko-desuka its 2am for me, im going to wrap it up for the night, laters

**Comment by doko-desuka** [comment]
Okay, later.

**Comment by sraedler** [comment]
Nice to see you working together :)

@doko-desuka where can I expect to see your final work? :)))

**Comment by doko-desuka** [comment]
@sraedler yes, I'm sure it's very convenient to you.

I haven't yet submitted an update to Arias800. Part of the Session subclass is this at the moment. I repeat that this is for video streaming on Kodi, you need to rewrite it for your own needs:  
```
 class CloudflareAdapter(requests.adapters.HTTPAdapter):
    # Cloudflare-friendly HTTP adapter based on VeNoMouS' Cloudscraper.
    # (1: https://github.com/VeNoMouS/cloudscraper/)
    # (2: https://www.npmjs.com/package/cloudscraper#recaptcha)
    def __init__(self, **kwargs):
        self.ssl_context = requests.packages.urllib3.util.ssl_.create_urllib3_context(
            ssl_version = ssl.PROTOCOL_SSLv23, # SSLv2, SSLv3, TLSv1, TLSv1.1, TLSv1.2
            cert_reqs = ssl.CERT_NONE,
            options = None, # Defaults to ssl.OP_NO_SSLv2, ssl.OP_NO_SSLv3 and ssl.OP_NO_COMPRESSION.
            ciphers = 'GREASE-2A2A:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA384:ECDHE+AESGCM:'
            'ECDHE+CHACHA20:DHE+AESGCM:DHE+CHACHA20:ECDH+AESGCM:DH+AESGCM:ECDH+AES:DH+AES:RSA+AESGCM:RSA+AES:'
            '!aNULL:!eNULL:!MD5:!DSS'
        )
        self.ssl_context.check_hostname = False
        super(CloudflareAdapter, self).__init__(**kwargs)

    def init_poolmanager(self, *args, **kwargs):
        kwargs['ssl_context'] = self.ssl_context
        return super(CloudflareAdapter, self).init_poolmanager(*args, **kwargs)

    def proxy_manager_for(self, *args, **kwargs):
        kwargs['ssl_context'] = self.ssl_context
        return super(CloudflareAdapter, self).proxy_manager_for(*args, **kwargs)
        
        
class CloudflareScraper(requests.Session):
    def __init__(self, *args, **kwargs):
        super(ScraperSession, self).__init__(*args, **kwargs)

        # Spoof a Chrome Android browser on a Nexus 7 device.
        self.mobileHeaders = OrderedDict((
            ('header', 'value'),
            (...)
        ))
        self.headers = self.mobileHeaders.copy()
        self.mount('https://', CloudflareAdapter())

        # Used in 'cfSolveChallenge()'.
        self.cfTestFlag = False
        
        
    (...)
        

    def cfSolveChallenge(self, resp, **kwargs):
        if self.cfTestFlag:
            self.cfTestFlag = False
            raise Exception('Repeated Cloudflare attempt')

        url = resp.url
        urlSplit = url.split('/', 3) # Splits 3 times, for 4 items: 'http(s):', '', 'domain.com', (path of the URL).
        domain = urlSplit[2]

        submitData = OrderedDict(kwargs.pop('data', {}))
        headers = kwargs.setdefault("headers", {})
        headers["Referer"] = url

        html = resp.text
        formIndex = html.find('<form id="challenge-form"')
        if formIndex == -1:
            raise Exception('Form not found')

        subHTML = html[formIndex:]
        submitURL = re.search('''action=['"]([^'"]+)''', subHTML).group(1)
        if submitURL.startswith('/'):
            submitURL = urlSplit[0] + '//' + domain + submitURL
        if not submitURL.startswith('http'):
            raise Exception('Unexpected form action found')

        submitData.update(
            match.groups()
            for match in re.finditer('''name=['"]([^'"]+).*?value=['"]([^'"]+)''', subHTML)
        )

        delayAmount = float(re.search('submit\(\).*?(\d+)', html, re.DOTALL).group(1)) / 1000.0

        if 'id="cf-dn-' in subHTML:
            extraDivExpression = re.search('id="cf-dn-.*?>(.*?)<', subHTML).group(1)

        # Initial value of the answer.
        jsAnswerValue = self.cfParseExpression(
            re.search('setTimeout\(function\(.*?:(.*?)}', html, re.DOTALL).group(1)
        )
        # Extract the arithmetic operations.
        builder = re.search("challenge-form'\);\s*;(.*);a.value", html, re.DOTALL).group(1)
        # Remove a function semicolon before splitting on semicolons, else it messes the order.
        lines = builder.replace(' return +(p)}();', '', 1).split(';')

        for line in lines:
            if len(line) and '=' in line:
                heading, expression = line.split('=', 1)

                if 'eval(eval(atob' in expression or 'eval(eval(e(' in expression:
                    # Uses the expression in an external <div>.
                    expressionValue = self.cfParseExpression(extraDivExpression)
                elif '(function(p' in expression:
                    # Expression + domain sampling function.
                    expressionValue = self.cfParseExpression(expression, domain)
                else:
                    expressionValue = self.cfParseExpression(expression)

                jsAnswerValue = self.cfArithmeticOp(heading[-1], jsAnswerValue, expressionValue)

        if '+ t.length' in html:
            jsAnswerValue += len(domain)

        submitData["jschl_answer"] = '%.10f' % jsAnswerValue

        time.sleep(delayAmount + 0.2)

        # Set a flag to avoid repeated attempts in case there's a wrong answer.
        self.cfTestFlag = True
        r = self.request('POST', submitURL, data=submitData, **kwargs)
        self.cfTestFlag = False
        return r 
```

The style is slightly different than the code already in Arias800 (it's using `camelCase` rather than `snake_case`), but the functions omitted stay the same, no change there other than their names.

**Comment by VeNoMouS** [comment]
Just as a heads up... I have written a native python implementation for the challenge solve , loosey based off all the other ones i've seen and am aware of

I'll do some more testing before I push it into dev, i could make this the default interpreter instead of js2py , removing the requirement for js2py in the requirements.txt

**Comment by VeNoMouS** [comment]
It's currently setting in dev branch

**Comment by sraedler** [comment]
Awesome!!!

I will test this with a few pages!

Is it possible to get rid of OpenSSL, since it is not default in Kodi? Or make it optional.

**Comment by VeNoMouS** [comment]
I'll see what I can do......

**Comment by VeNoMouS** [comment]
One of my testers did find a site the regex broke on, so i need to tune it a little more.

**Comment by sraedler** [comment]
I also had the problem with broken regex

**Comment by VeNoMouS** [comment]
I need to dive into this a little more work out the problem is, as there is more to it than simply fixing the regex in this case.

**Comment by doko-desuka** [comment]
@VeNoMouS hi, I'll take a look and report back.

**Comment by doko-desuka** [comment]
I couldn't find what's wrong with the native version. It does make a different jschl_answer value than what my browser is making.  

I thought of this other way to split the challenge into all the operations, rather than using the semicolons and having to replace that function:
```
        jsfuckChallenge = re.search(
            r"setTimeout\(function\(\){\s+var.*?f,\s*(?P<var>\w+).*?:(?P<init>\S+)};.*"
            r"?\(\'challenge-form\'\);\s*;(?P<challenge>.*?a\.value)"
            r"(?:.*id=\"cf-dn-.*?>(?P<k>\S+)<)?",
            body,
            re.DOTALL | re.MULTILINE
        ).groupdict()

        jsfuckChallenge['challenge'] = re.findall(
            jsfuckChallenge['var']+'.*?([+\-*/])=(.*?);(?=a\.value|'+jsfuckChallenge['var']+')',
            jsfuckChallenge['challenge']
        )

(...)

        # ------------------------------------------------------------------------------- #

        for oper, expression in jsfuckChallenge['challenge']: # Unpacks into the operation and expression already.
```
But it doesn't alter the result, it just uses regex instead of string functions.

**Comment by VeNoMouS** [comment]
What i found when i changed `k` variable's search..

To the following
```python
r"(.*?\"\s+id=\"cf\-.*?\">(?P<k>\S+)</|)"
```
it generated the following answer..
```
372.2126298751
```

instead of 

```391.2126298751```

as we can see in the following..

```python
>>> from lib import cloudscraper
>>> with open('/tmp/body4.html', 'r') as oFile:
...    data = oFile.read()
...
>>> scraper.IUAM_Challenge_Response(data, 'https://www.cornerstreet.fr/catalog/product/view/id/150417/category/445/', 'native')
>>> cloudscraper.create_scraper().IUAM_Challenge_Response(data, 'https://www.cornerstreet.fr/catalog/product/view/id/150417/category/445/', 'native')
{'url': 'https://www.cornerstreet.fr/catalog/product/view/id/150417/category/445/?__cf_chl_jschl_tk__=2b5411cb82763f45fda2941c8ab9239470463a4b-1575104305-0-ASYXkeTmSIu26ytbuRGWh3xlf8hBvWEZ_P5hlB-ebt9Uv2u2BZuZULNS579tVuU0vLEAVzDa2p2spDg209_O2Dw2C-pSTyhv7xFyvGVyKGytOCOmJrE49ulnbAymDtERNwIN8Osv1pmz_xXCJaXjrzq84jcRhdSIM1jgiK5_wlY3H4pJeGHFKj7-ge-jm_rz_rIah5JUVLeJZyrMd7R7QgzZKtZtSaeydDaML11RCV7h2mKlQAC9Ye5MfO1TgeXZalbgGYwG9JOiqDOAIabajOSt-fJMc37SsO0xPUhMKNPclGSgnviDdv5yZDtHxujdudq18dDVKQ58GbRGnjJ61fvOz55WAOq9EXCFOTrm3Ehsd9niklvodqo631ekJ69j_w', 'data': OrderedDict([('r', '1a92c46efb29f3022024f5d52c1f0ec7996620a0-1575104305-0-AW5TECtQdpbmjg2klkZmcOUYiysUP4+tj2r1T8tmK8hRIgnQf0fRydF1hbgbJsnO6kfQeq0Ej8MYTLnKKJw+Efh1puJmnTx0cafpCTiqI9UOWxB3s66d+sQvqFwyqWjafA0Mq1zI9QAvnocPDErxCLMLZBYM4DmDKM8lkZIeL2EyEBzgjXlC0Au3IYAqIe/9AfXVeC8dGGf4Iv7pK4vVAQ5tQh62eVnHPrU7eK8wkAwJKKvZ4AQceYNW1uwyAroBZ6vpz6s4GiLo3wVMtlVbUL59qfjRSWAGJ5wUbZfXvIo3g9xHo4flLaf8h6H2+Z0vchCCIVgmYIzm9WidrGjoSLFB6kbyVJAsUh8wTRHQ1dZcnkWhh0KMCXEEglCxrccImJO99XNN6ctU9CECuClMQtRYkjjEciT/3WE8zudIAeaRsPHSQ94DADpKafaVzS7CLio/IZkV1ahsxsUWX1Km9A7KOsl72tThnvIR0KVqdgiIavxwnmwyMtvbZ2cqdKecZQGi9+0E6jFpskM7lzvNV8nj6gIYCYntAIL/sQWVjKjSH18wjWdwrdCC2qPkfeQmqzOcs0Q26llR+McjR4I6RucQA+EZfpXBOkT0RE25ilVRSmSQCSobEdg+0voiWSwRTK6c3IxBehnTndGIn3uySvy1XVABf2NMl/OJQ40XlT4Y/UrI3kShcjbbRDA4aryDyjNNS7ecNaTs+bqFA4BM3e5yxl4GWm/GtcQx9oGM3FKqYjaUtIQoVEGlUO6Hl9Johq+eLQfPlKNzBxN8XGLogpCwaayfCQaffsMvNWa6HF8Wv9OhZGQBvAC56uqyG1yjI0FVuFYiNR14dhyj2DJmKtG68dq7vmQ1QKgxE1RlBZrKR4NZQreF2GmKbmU2bep7c8xRg73OYyCkPkpTZ5x8gxMocp7ElWWxR/3U02GY/aZgvTRYdmVwv/w7/St61lokmA7hDnQZx8N9CuTlLtYpEUJsxQta93bobwd93TsrB7g6fvkoGTR0vatH0mFe4sWlD7kfa/WsSS64Yk9PL1L6hEcgxZB1yeIJXRjN7VT6JFGy3kqEOAoEGzMbnA6hSglvExo4nQ2F1/aTNOC4UCjSnFK5Xzwu+czMuqr2Yg5rs9wG6I7XfxwPVExRnawjVs4DCRo3OSb6wyapQVbZI+bf7ECK6vJGnxTyWx38YipmuIqtvl8mX96xsVSxhcswShv3Dy6O1fwp4mfqoV3vwFdaglCmqvOkwahR1YxMucDW8/VDSKeEgPAn9rFzbn2qghSfHDTcoaXXos+b1HYSTILA37i3b9C9YBzLwHiZAm/esjoW'), ('jschl_vc', 'e031d011810caae25fa60d66c1f0ea81'), ('pass', '1575104309.002-x4PTv9bJV1'), ('jschl_answer', 372.2126298751)])}
>>> cloudscraper.create_scraper().IUAM_Challenge_Response(data, 'https://www.cornerstreet.fr/catalog/product/view/id/150417/category/445/', 'js2py')
{'url': 'https://www.cornerstreet.fr/catalog/product/view/id/150417/category/445/?__cf_chl_jschl_tk__=2b5411cb82763f45fda2941c8ab9239470463a4b-1575104305-0-ASYXkeTmSIu26ytbuRGWh3xlf8hBvWEZ_P5hlB-ebt9Uv2u2BZuZULNS579tVuU0vLEAVzDa2p2spDg209_O2Dw2C-pSTyhv7xFyvGVyKGytOCOmJrE49ulnbAymDtERNwIN8Osv1pmz_xXCJaXjrzq84jcRhdSIM1jgiK5_wlY3H4pJeGHFKj7-ge-jm_rz_rIah5JUVLeJZyrMd7R7QgzZKtZtSaeydDaML11RCV7h2mKlQAC9Ye5MfO1TgeXZalbgGYwG9JOiqDOAIabajOSt-fJMc37SsO0xPUhMKNPclGSgnviDdv5yZDtHxujdudq18dDVKQ58GbRGnjJ61fvOz55WAOq9EXCFOTrm3Ehsd9niklvodqo631ekJ69j_w', 'data': OrderedDict([('r', '1a92c46efb29f3022024f5d52c1f0ec7996620a0-1575104305-0-AW5TECtQdpbmjg2klkZmcOUYiysUP4+tj2r1T8tmK8hRIgnQf0fRydF1hbgbJsnO6kfQeq0Ej8MYTLnKKJw+Efh1puJmnTx0cafpCTiqI9UOWxB3s66d+sQvqFwyqWjafA0Mq1zI9QAvnocPDErxCLMLZBYM4DmDKM8lkZIeL2EyEBzgjXlC0Au3IYAqIe/9AfXVeC8dGGf4Iv7pK4vVAQ5tQh62eVnHPrU7eK8wkAwJKKvZ4AQceYNW1uwyAroBZ6vpz6s4GiLo3wVMtlVbUL59qfjRSWAGJ5wUbZfXvIo3g9xHo4flLaf8h6H2+Z0vchCCIVgmYIzm9WidrGjoSLFB6kbyVJAsUh8wTRHQ1dZcnkWhh0KMCXEEglCxrccImJO99XNN6ctU9CECuClMQtRYkjjEciT/3WE8zudIAeaRsPHSQ94DADpKafaVzS7CLio/IZkV1ahsxsUWX1Km9A7KOsl72tThnvIR0KVqdgiIavxwnmwyMtvbZ2cqdKecZQGi9+0E6jFpskM7lzvNV8nj6gIYCYntAIL/sQWVjKjSH18wjWdwrdCC2qPkfeQmqzOcs0Q26llR+McjR4I6RucQA+EZfpXBOkT0RE25ilVRSmSQCSobEdg+0voiWSwRTK6c3IxBehnTndGIn3uySvy1XVABf2NMl/OJQ40XlT4Y/UrI3kShcjbbRDA4aryDyjNNS7ecNaTs+bqFA4BM3e5yxl4GWm/GtcQx9oGM3FKqYjaUtIQoVEGlUO6Hl9Johq+eLQfPlKNzBxN8XGLogpCwaayfCQaffsMvNWa6HF8Wv9OhZGQBvAC56uqyG1yjI0FVuFYiNR14dhyj2DJmKtG68dq7vmQ1QKgxE1RlBZrKR4NZQreF2GmKbmU2bep7c8xRg73OYyCkPkpTZ5x8gxMocp7ElWWxR/3U02GY/aZgvTRYdmVwv/w7/St61lokmA7hDnQZx8N9CuTlLtYpEUJsxQta93bobwd93TsrB7g6fvkoGTR0vatH0mFe4sWlD7kfa/WsSS64Yk9PL1L6hEcgxZB1yeIJXRjN7VT6JFGy3kqEOAoEGzMbnA6hSglvExo4nQ2F1/aTNOC4UCjSnFK5Xzwu+czMuqr2Yg5rs9wG6I7XfxwPVExRnawjVs4DCRo3OSb6wyapQVbZI+bf7ECK6vJGnxTyWx38YipmuIqtvl8mX96xsVSxhcswShv3Dy6O1fwp4mfqoV3vwFdaglCmqvOkwahR1YxMucDW8/VDSKeEgPAn9rFzbn2qghSfHDTcoaXXos+b1HYSTILA37i3b9C9YBzLwHiZAm/esjoW'), ('jschl_vc', 'e031d011810caae25fa60d66c1f0ea81'), ('pass', '1575104309.002-x4PTv9bJV1'), ('jschl_answer', 391.2126298751)])}
>>> cloudscraper.create_scraper().IUAM_Challenge_Response(data, 'https://www.cornerstreet.fr/catalog/product/view/id/150417/category/445/', 'nodejs')
{'url': 'https://www.cornerstreet.fr/catalog/product/view/id/150417/category/445/?__cf_chl_jschl_tk__=2b5411cb82763f45fda2941c8ab9239470463a4b-1575104305-0-ASYXkeTmSIu26ytbuRGWh3xlf8hBvWEZ_P5hlB-ebt9Uv2u2BZuZULNS579tVuU0vLEAVzDa2p2spDg209_O2Dw2C-pSTyhv7xFyvGVyKGytOCOmJrE49ulnbAymDtERNwIN8Osv1pmz_xXCJaXjrzq84jcRhdSIM1jgiK5_wlY3H4pJeGHFKj7-ge-jm_rz_rIah5JUVLeJZyrMd7R7QgzZKtZtSaeydDaML11RCV7h2mKlQAC9Ye5MfO1TgeXZalbgGYwG9JOiqDOAIabajOSt-fJMc37SsO0xPUhMKNPclGSgnviDdv5yZDtHxujdudq18dDVKQ58GbRGnjJ61fvOz55WAOq9EXCFOTrm3Ehsd9niklvodqo631ekJ69j_w', 'data': OrderedDict([('r', '1a92c46efb29f3022024f5d52c1f0ec7996620a0-1575104305-0-AW5TECtQdpbmjg2klkZmcOUYiysUP4+tj2r1T8tmK8hRIgnQf0fRydF1hbgbJsnO6kfQeq0Ej8MYTLnKKJw+Efh1puJmnTx0cafpCTiqI9UOWxB3s66d+sQvqFwyqWjafA0Mq1zI9QAvnocPDErxCLMLZBYM4DmDKM8lkZIeL2EyEBzgjXlC0Au3IYAqIe/9AfXVeC8dGGf4Iv7pK4vVAQ5tQh62eVnHPrU7eK8wkAwJKKvZ4AQceYNW1uwyAroBZ6vpz6s4GiLo3wVMtlVbUL59qfjRSWAGJ5wUbZfXvIo3g9xHo4flLaf8h6H2+Z0vchCCIVgmYIzm9WidrGjoSLFB6kbyVJAsUh8wTRHQ1dZcnkWhh0KMCXEEglCxrccImJO99XNN6ctU9CECuClMQtRYkjjEciT/3WE8zudIAeaRsPHSQ94DADpKafaVzS7CLio/IZkV1ahsxsUWX1Km9A7KOsl72tThnvIR0KVqdgiIavxwnmwyMtvbZ2cqdKecZQGi9+0E6jFpskM7lzvNV8nj6gIYCYntAIL/sQWVjKjSH18wjWdwrdCC2qPkfeQmqzOcs0Q26llR+McjR4I6RucQA+EZfpXBOkT0RE25ilVRSmSQCSobEdg+0voiWSwRTK6c3IxBehnTndGIn3uySvy1XVABf2NMl/OJQ40XlT4Y/UrI3kShcjbbRDA4aryDyjNNS7ecNaTs+bqFA4BM3e5yxl4GWm/GtcQx9oGM3FKqYjaUtIQoVEGlUO6Hl9Johq+eLQfPlKNzBxN8XGLogpCwaayfCQaffsMvNWa6HF8Wv9OhZGQBvAC56uqyG1yjI0FVuFYiNR14dhyj2DJmKtG68dq7vmQ1QKgxE1RlBZrKR4NZQreF2GmKbmU2bep7c8xRg73OYyCkPkpTZ5x8gxMocp7ElWWxR/3U02GY/aZgvTRYdmVwv/w7/St61lokmA7hDnQZx8N9CuTlLtYpEUJsxQta93bobwd93TsrB7g6fvkoGTR0vatH0mFe4sWlD7kfa/WsSS64Yk9PL1L6hEcgxZB1yeIJXRjN7VT6JFGy3kqEOAoEGzMbnA6hSglvExo4nQ2F1/aTNOC4UCjSnFK5Xzwu+czMuqr2Yg5rs9wG6I7XfxwPVExRnawjVs4DCRo3OSb6wyapQVbZI+bf7ECK6vJGnxTyWx38YipmuIqtvl8mX96xsVSxhcswShv3Dy6O1fwp4mfqoV3vwFdaglCmqvOkwahR1YxMucDW8/VDSKeEgPAn9rFzbn2qghSfHDTcoaXXos+b1HYSTILA37i3b9C9YBzLwHiZAm/esjoW'), ('jschl_vc', 'e031d011810caae25fa60d66c1f0ea81'), ('pass', '1575104309.002-x4PTv9bJV1'), ('jschl_answer', 391.2126298751)])}
>>>
```

Let me test your above fix

**Comment by VeNoMouS** [comment]
NM worked it out... t.length .... add domain length to the end variable..

**Comment by VeNoMouS** [comment]
Also... @doko-desuka , nice job on the regex change... in regards to doing it via regex over strings :)

**Comment by VeNoMouS** [comment]
@doko-desuka 

```python
    jsfuckChallenge = re.search(
        r"setTimeout\(function\(\){\s+var.*?f,\s*(?P<variable>\w+).*?:(?P<init>\S+)};"
        r".*?\(\'challenge-form\'\);\s+;(?P<challenge>.*?a\.value)"
        r"(?:.*id=\"cf-dn-.*?>(?P<k>\S+)<)?",
        body,
        re.DOTALL | re.MULTILINE
    ).groupdict()

    jsfuckChallenge['challenge'] = re.findall(
        '{}.*?([+\-*/])=(.*?);(?=a\.value|{})'.format(
            jsfuckChallenge['variable'],
            jsfuckChallenge['variable']
        ),
        jsfuckChallenge['challenge']
    )
```
cleaned your regex up a bit, but i really like where you went with that, cheers

**Comment by doko-desuka** [comment]
> if not jsfuckChallenge.get('k') and 't.length' in body:
>    jschl_answer += len(domain)

Nice catch! I forgot about that.  
It only appears on some challenges.  
I don't think you need to check for 'k' existence, only if '+ t.length' is in the content (I don't remember there being a case where 'k' wasn't present and t.length was, but it should be safer anyway).

**Comment by devilsport** [comment]
Great work! Compliments to everyone :) I tested current dev branch in Kodi and works perfectly. I just had to disable the OpenSSL import and use of OpenSSL.SSL.Error exception.

**Comment by reddit-reaper** [comment]
well this was unexpected lol thanks @VeNoMouS the devs that work on stuff in A4K-OpenProject had just updated our cfscrape module to solve for new challenges and this makes it so they no longer have to worry about it anymore :) if you want, i extend an open invitation to our dev group on keybase. if you message me on keybase username: redditreapercody i will get you into the group. devs would love to chat with you if possible :)

**Comment by VeNoMouS** [comment]
Re OpenSSL .. from memory i had to use it because of some wintendo systems... I will run some tests and see if i can get away without it on all systems

**Comment by doko-desuka** [comment]
The standard `ssl` module imports fine in my Android TV box, so there's this as fallback.  
If the external library is Python-only then one'd need to pack it into a script.module.[addon_name] type of add-on for importing into other add-ons after you add it as a dependency in the [addon.xml](https://kodi.wiki/view/Addon.xml#.3Cimport.3E) metadata file. If the external library has binaries, they'd need to be built for each Kodi platform and put in script.module.\* add-ons in the same way.
More details in here:
- https://forum.kodi.tv/showthread.php?tid=231642&pid=2051095#pid2051095
- https://forum.kodi.tv/showthread.php?tid=251555

The `plugin.video.elementum` add-on is probably the best example of this, has binaries for lots of different platforms.  
- https://github.com/elgatito/plugin.video.elementum#build

**Comment by sraedler** [comment]
@doko-desuka that's true, but additional overhead for both deploying a new version and for the performance of cloudscraper.

@VeNoMouS  I also want to ask whether it would make sense to move some imports to the lines, where they are needed. Since for example recaptcha is mostly not used. Or is this just too less improvement compared to readability?

**Comment by VeNoMouS** [comment]
@sraedler what the hell are you on about? reCaptcha is dynamically imported at the time its loaded.

**Comment by VeNoMouS** [comment]
OK re the reason OpenSSL for exceptions.. is because of urllib3 which is used by requests, it was fixed in February of this year, however.... we are only using them in the `loadCipherSuite` function, And I was looking at doing away with that, so i'll remove it out over the next day or so.

**Comment by sraedler** [comment]
Sorry I tried to implement a my.JDownloader extension for recaptcha solving which was quite slow on import. And for testing purpose I just replaced the abstract class with my real implementation and that's why I thought the import is slow. My fault! :zipper_mouth_face:

**Comment by VeNoMouS** [comment]
As promised I have removed the requirement for OpenSSL import, current in dev branch with commit 5c2901a0c32e21974c693fcf3d5fc59574c4b2f6

**Comment by VeNoMouS** [comment]
Should all be resolved in release 1.2.16 package.
